### ISSUE TITLE
Investigate Applying PowerSGD to Communication Hook for Gradient Compression

### ISSUE THREAD

**Issue by SciPioneer** [pytorch/pytorch#47202]
## 🚀 Feature
PowerSGD can be potentially used for gradient compression: https://arxiv.org/abs/1905.13727. Investigate this algorithm in the context of communication hook.

## Motivation

PowerSGD can still keep the associativity/linearity of the gradients after compression, and hence it can still be implemented efficiently by using the native communication library like NCCL. It can compress every tensor of size M*N that represents a variable into 2 smaller tensors M * rank and N * rank for communication. Note that 3D or higher-rank tensors can also be supported, and the compression ratio can be computed by viewing the higher-rank tensor as a 2D tensor.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @xush6528

**Comment by tvogels** [comment]
I'd be happy to help out with this investigation and very curious about your findings!

**Comment by SciPioneer** [comment]
> If you need any support, I'd be happy to help out with this investigation and very curious about your findings!

Could you please take a look at #48060 (particularly, file powerSGD_hook.py). It implements a simplified version, as the input of the compression hook is a large flattened tensor that batches a tensor list (mixed 1D and multi-dimensional tensors.) Thanks!

**Comment by SciPioneer** [comment]
@tvogels 

Hi Thijs, could you please take a look at #48670 (mainly powerSGD_hook method in the file powerSGD_hook.py)? This PR adds error feedback.

As I understand, the error feedback implementation should be simple for my case, which can be viewed as vanilla SGD, as my comm hook module is not fused with optimizer, and I don't need to follow the example in PowerSGD paper and consider momentum.

The initial error tensor is all 0s at the first iteration, in the same shape as the input grad tensor. In every iteration, the extra changes are:
1) Update the input gradient by adding the local error at the beginning;
2) Keep a copy of this gradient before applying PowerSGD;
3) Update the local error as the diff between the synced gradient and the copy created at step 2). At this moment, P and Q are already all-reduced, and the decompressed gradient is computed based on the synced P and Q.

I used a dict that maps every tensor/matrix to a local error in the state of this compression hook, so all the local errors are stored in the PowerSGD state and not paired up with the input gradients by default. This can avoid the extra memory costs if PowerSGD is not enabled.

I have a follow-up PR to fix the key in error_dict, and this bug actually is not necessary exposed, as long as the buckets are rebuilt in the training. However, I find that the error feedback is still buggy, but not sure what's wrong.

Thanks!

**Comment by SciPioneer** [comment]
> If you need any support, I'd be happy to help out with this investigation and very curious about your findings!

Hi Thijs! Thanks for the help of code review! An initial version of PowerSGD communication hook has been checked in, and it has shown some early promising results. I am running more experimental evaluations, but this also requires more time on our end. Could you please help also compare it with your original implementation? Hopefully the performance can be on par with your implementation.

Just need a few lines of change to apply PowerSGD compression on PyTorch, given a DDP model instance:

```
import torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook as powerSGD

        # Assume a ddp_model is created here.
        state = powerSGD.PowerSGDState(
            process_group=process_group, matrix_approximation_rank=1
        )
        ddp_model.register_comm_hook(state, powerSGD.powerSGD_hook)
```

Thanks!

**Comment by SciPioneer** [comment]
**Summary:** 

After extensive evaluation of PowerSGD communication hook on multiple benchmarks, we have seen some speedups without accuracy loss for some use cases. As a result, we have added this feature to our next release 1.8 as a BETA feature, and there is a dedicated [documentation page](https://github.com/pytorch/pytorch/blob/master/docs/source/ddp_comm_hooks.rst) on PyTorch Tutorial to introduce DDP communication hooks including this PowerSGD hook.

**Highlights**

1. Since we have implemented PowerSGD as a communication hook, it is very easy to use, just like a plug-in.
2. As validated by PowerSGD paper author @tvogels, the performance of PowerSGD communication hook is on par with the implementation in the paper, in terms of both training speed and accuracy.
2. To better balance the trade-off between speedup and accuracy, our implementation introduces another algorithm hyperparameter `start_powerSGD_iter` to defer PowerSGD compression.
3. We also provide a variant called `batched_powerSGD_hook`, which may be more effective when `matrix_approximation_rank`=1.

**When Shall I Consider PowerSGD?**

- AllReduce communication is the major performance bottleneck.

Usually this means that 1) the model has a large number of parameters in the dense layers, and 2) the communication is not largely overlapped with computation. Therefore, before trying powerSGD, users can first check the model architecture and use libgpumon profiler to measure the overlap between AllReduce and computation. 

- The gradient tensors can be well approximated by low-rank tensors.

Although it is very hard to know whether this premise can hold given a model, now the ease of use of PowerSGD communication hook can make the validation much easier.
